### ISSUE TITLE
Memory usage optimization

### ISSUE THREAD

**Issue by chenguoguo** [athena-team/athena#74]
I had a discussion with @tjadamlee , we think our default examples' GPU memory usage is way too high, which is blocking a lot of users. We should tune the parameters (e.g., batch size) as well as the model structures, to keep the memory usage under 8G for default example setups.

@Some-random could you please take the lead and cut all examples' memory usage under 8G, while keeping the performance as much as possible?

**Comment by Some-random** [comment]
For model structures, I've tried changing d_model from 512 to 256, which almost cut trainable parameters to half (and WER on HKUST increases from 23.3 to 23.55), but the GPU memory usage is still over 17G, which is really confusing...

Decreasing batch size will normally hurt the performance of Transformer models. I guess we could  keep the 'effective batch size' unchanged by applying the accum_grads technique, where we use a smaller batch size (say 16) and update parameters once every two batches (for the batch that doesn't update parameter, we just calculate its gradients and store it). By doing so we can still enjoy the benefits of big batch size and constrain GPU memory usage.

**Comment by chenguoguo** [comment]
Did you try per_process_gpu_memory_fraction and allow_growth?

**Comment by iou2much** [comment]
I try to use mix-precision, which can speed up a lot for MPC stage. but I got nan loss in fine-tuning stage. could anyone help please? I describe the situation here: #77

**Comment by tjadamlee** [comment]
@Some-random we may build another new example which can be trained using a smaller memory and do not focus on sota

**Comment by Some-random** [comment]
> @Some-random we may build another new example which can be trained using a smaller memory and do not focus on sota

I've created a pull request with a new demo conf https://github.com/athena-team/athena/pull/78.

If it looks good to you, I'll mention this new conf in README.md.

**Comment by Some-random** [comment]
> Did you try per_process_gpu_memory_fraction and allow_growth?

Allow_growth is enabled by default. I tried per_process_gpu_memory_fraction and it manages to keep GPU usage constant but it also makes running really slow (4x time on one epoch)...

**Comment by Some-random** [comment]
Closing this issue for now. Please reopen it if there exists further question.
